File: week3/D15_1GenFact.hs
import Data.Function ((&))
import Data.Bits ((.&.))

generator modulus factor = flip mod modulus . (*) factor

genA = generator 2147483647 16807
genB = generator 2147483647 48271

judge startA startB =
    let
        zipAB (a,b) = (genA a, genB b)
        numberStream = iterate zipAB (startA, startB)
        suffixEqual (x,y) = if (0xffff .&. x) == (0xffff .&. y) then 1 else 0
        fourtyMillion = 4 * 10^7
    in
        map suffixEqual numberStream
        & take fourtyMillion
        & sum

readStarts :: String -> (Int, Int)
readStarts = (\[sa,sb] -> (sa, sb)) . map (read . last . words) . lines

main = interact (show . uncurry judge . readStarts)

File: complement/D15_1GenFact.rs
fn judge(start_a: usize, start_b: usize) -> usize {
    let gen_a = |a| (a * 16807) % 2147483647;
    let gen_b = |b| (b * 48271) % 2147483647;
    let mut a = gen_a(start_a);
    let mut b = gen_b(start_b);
    let mut sum = 0;
    for _ in 0..40_000_000 {
        if (a & 0xffff) == (b & 0xffff) { sum += 1 }
        a = gen_a(a); b = gen_b(b);
    }
    sum
}

fn main() { println!("{}", judge(65, 8921)) }
